flink/src/main/java/org/apache/zeppelin/flink/sql/TimeSeriesStreamSqlJob.java

```diff
@@ -0,0 +1,144 @@
+  }
+
+  @Override
+  protected String getType() {
+    return "ts";
+  }
+
+  @Override
+  protected List<String> getValidLocalProperties() {
+    return Lists.newArrayList("type", "parallelism",
+            "refreshInterval", "enableSavePoint", "runWithSavePoint", "threshold");
+  }
+
+  @Override
+  protected void checkTableSchema(TableSchema schema) throws Exception {
+    //    if (!(schema.getFieldDataType(0).get() instanceof TimestampType)) {
+    //      throw new Exception("The first column must be TimestampType, but is " +
+    //              schema.getFieldDataType(0));
+    //    }
+  }
+
+  @Override
+  protected void processInsert(Row row) {
+    LOGGER.debug("processInsert: " + row.toString());
+    materializedTable.add(row);
+  }
+
+  @Override
+  protected void processDelete(Row row) {
+    throw new RuntimeException("Delete operation is not expected");
+  }
+
+  @Override
+  protected void refresh(InterpreterContext context) {
+    context.out().clear();
+    try {
+      context.out.write("%table(");
+      context.out.write("type=ts,threshold=" + tsWindowThreshold);
+      context.out.write(",columns=");
+      for (int i = 0; i < schema.getFieldCount(); ++i) {
+        String field = schema.getFieldNames()[i];
+        context.out.write(field);
+        if (i != (schema.getFieldCount() - 1)) {
+          context.out.write(":");
+        }
+      }
+      context.out.write(")\n");
+
+      for (int i = 0; i < schema.getFieldCount(); ++i) {
+        String field = schema.getFieldNames()[i];
+        context.out.write(field);
+        if (i != (schema.getFieldCount() - 1)) {
+          context.out.write("\t");
+        }
+      }
+      context.out.write("\n");
+
+      // sort it by the first column
+      materializedTable.sort((r1, r2) -> {
+        String f1 = r1.getField(0).toString();
+        String f2 = r2.getField(0).toString();
+        return f1.compareTo(f2);
+      });
+
+      if (materializedTable.size() != 0) {
+        long maxTimestamp =
+                ((java.sql.Timestamp) materializedTable.get(materializedTable.size() - 1)
+                .getField(0)).getTime();
+
+        materializedTable = materializedTable.stream()
+                .filter(row -> ((java.sql.Timestamp) row.getField(0)).getTime() >
+                        maxTimestamp - tsWindowThreshold)
+                .collect(Collectors.toList());
+
+        LOGGER.debug("*****************Row size: " + materializedTable.size());
+
+        for (Row row : materializedTable) {
+          for (int i = 0; i < row.getArity(); ++i) {
+            Object field = row.getField(i);
+            context.out.write(field.toString());
+            if (i != (row.getArity() - 1)) {
+              context.out.write("\t");
+            }
+          }
+          LOGGER.debug("Row:" + row);
+          context.out.write("\n");
+        }
+      }
+      context.out.flush();
+    } catch (IOException e) {
+      e.printStackTrace();
```

Comment:
remove?

there seems to be some duplicated code here?

Reply:
fixed